content/recipes/authoring/kata-snippets/submission-tests.md

```diff
@@ -0,0 +1,88 @@
+- **Tests should handle unexpected answers gracefully.** Tests should not crash, and should present meaningful error messages when an unexpected answer is returned by the user's solution, e.g. `null` references when none is expected or an object of the incorrect type.
+- **Tests should not print excessively to standard output.** Console output is limited to 1.5 MB and when the limit is reached, the test suite is aborted by the runner and marked as failed. Users sometimes debug their solutions by printing to stdout, and when the test suite itself uses up a significant fraction of the buffer, users may not have enough left for debugging. This especially applies to kata which output HTML to create some visualizations: such HTML uses up a lot of the buffer space allowed for stdout. Ideally, the test suite should give detailed output **only when a test fails**. Successful tests should be as quiet as possible.
+- **Provide both [fixed](#fixed-tests) and [random](#random-tests) tests** for the submit tests. Some test frameworks offer features like test case generators (`QuickCheck`, `ScalaCheck`), and when used, authors can choose different structures for their test suite, but they should still create test specifications covering all possible scenarios. See specific sections below for more details.
+- **Tests should not be susceptible to input mutation.** It should not be possible for the user's code to modify its input in a way which would influence the execution, result, or output of a test. If a reference solution is used, it should be called first, and the user solution should be called afterwards with the same input, or only pass a (deep) copy of the inputs to the user's solution. Potentially mutated input should not be used to compose any information presented to the user.
+- **Assertions should be placed near invocations of the user's solution**, preferably immediately following them. When several results are computed calling the user's function but the related assertions are only done further, what is printed to the console by the user isn't close to the related assertion in the output panel anymore. Debugging becomes difficult for the user and that makes the experience unpleasant.
+- **The test suite should not take a long time to run.** For kata which are not performance-oriented, the test suite should complete well below the time limit, ideally within 3 seconds. There's a time limit of 12 seconds (16 or 20 seconds for a few languages) for every attempt, but the generation and execution of test cases, including any calls to the user and reference solutions, should run well below this limit. Otherwise, there is a risk that tests will inconsistently time out, or some valid, but not optimally performant solutions will be rejected.
```

Comment:
```suggestion
- **The test suite should not take a long time to run.** For kata which are not performance-oriented, the test suite should complete well below the time limit, ideally within 3 seconds. There's a time out limit of 12 seconds (16 or 20 seconds for a few languages) for every attempt, but the generation and execution of test cases, including any calls to the user and reference solutions, should run well below this limit. Otherwise, there is a risk that tests will inconsistently time out, or some valid, but not optimally performant solutions will be rejected for no objective reasons.
```

Reply:
Fixed. But do we really need that "for no objective reason" in the end?

Reply:
Isn't "a timeout" already "a time limit", which makes a "a timeout limit" a limit imposed on the timeout?

Reply:
no idea...

Reply:
removed `for no objective reason`. Still no idea bout `timeout limit`.

Reply:
I agree that we should just use "timeout" or keep "time limit".

Reply:
Leaving it as it is, then